Support for fixed[x]/pattern[x] via x = y (exactly)

cmoesel opened the pull request:
This PR supports the new syntax to allow authors to choose if they want to fix by `pattern[x]` or by `fixed[x]`:
```
* foo = "bar"           // uses pattern[x]
* foo = "bar" (exactly) // uses fixed[x]
```

This PR includes the following changes:
* Added support in the grammar and importer for `(exactly)`
* Updated `FixedValueRule` to have an `exactly` flag
* Added `toString()` methods to some FSH classes (returns FSH syntax)
* Added `toFHIR[X]()` methods to some FSH classes (where `[X]` is a FHIR type; e.g., `toFHIRQuantity()`)
* Now emits an error if an author attempts to override a `fixed[x]` with a `pattern[x]`
* Now checks for parent `pattern[x]`/`fixed[x]` whenever fixing any values
* Adds a `fixFHIRValue` function that replaces a lot of previous special case code
* Adds a bunch o' tests

I also ran this against `fsh-mcode` and compared the differences against master.  The only differences are that the examples now contain all of the `display` properties for their codes (which is expected, and good).

Fixes #218 and #187

cmoesel commented:
So it looks like I also inadvertently addressed another story in this (that I forgot was a separate story).  So this also addresses _CIMPL-291: FSH: Display name in fixed quantity should go in Quantity.unit_.

See:
* https://github.com/FHIR/sushi/blob/55b3b6250458968815beadbb2dea2ec19c67f6d2/src/fshtypes/FshCode.ts#L46-L48
* https://github.com/FHIR/sushi/blob/55b3b6250458968815beadbb2dea2ec19c67f6d2/src/fshtypes/FshQuantity.ts#L27-L29
* https://github.com/FHIR/sushi/blob/55b3b6250458968815beadbb2dea2ec19c67f6d2/test/fhirtypes/ElementDefinition.fixFshCode.test.ts#L434-L454

ngfreiter commented on src/errors/FixedToPatternError.ts:
```diff
@@ -0,0 +1,13 @@
+  ];
+  constructor(public fixedProperty: string) {
+    super(
+      `Cannot fix this element using a pattern; as it is already fixed in the StructureDefinition using ${fixedProperty}.`
```
It might not be immediately clear to someone why they can't fix with pattern if something is already fixed with fixed. Maybe this could include something about how pattern is a looser constraint than fixed.

cmoesel commented on src/fhirtypes/ElementDefinition.ts:
```diff
@@ -860,67 +884,127 @@ export class ElementDefinition {
+    let type: string = typeof value;
+    if (type === 'object' && value?.constructor?.name) {
+      // For types like FshCode, FshQuantity, etc, remove the Fsh so we're left with the real type name
+      type = value.constructor.name.replace(/^Fsh/, '');
```
So far we have.  The alternative approach would be to add a new read-only property on the class that indicates its FHIR type (or to have a global map for that -- but then we need to remember to update the map whenever we add a new FSH type).  Being able to get the intended type from the class somehow, however, is important to the approach that allows us to generalize a lot of the previously special-case code.

cmoesel replied:
Would you prefer a read-only property on the class?

ngfreiter replied:
I'd prefer a read only property.

cmoesel commented:
Thanks for the review @ngfreiter.  I'm going to wait for a second review before addressing the comments, as I want to just knock them all out at once.

cmoesel commented:
OK.  Comments addressed in 3ab610a.
